src/model.rs

```diff
@@ -787,88 +792,79 @@ impl Collections {
+            let last_stop = &prev_vj.stop_times.last();
+            let first_stop = &next_vj.stop_times.first();
+            match (last_stop, first_stop) {
+                (Some(last_stop), Some(first_stop)) => {
+                    if last_stop.stop_point_idx == first_stop.stop_point_idx {
+                        // We can discard when the stop points are identicals (see
+                        // Example 1 above).
+                        false
+                    } else {
+                        let prev_cal = self.calendars.get(&prev_vj.service_id);
+                        let next_cal = self.calendars.get(&next_vj.service_id);
+                        // for the stay in to be possible the vj should have at least one date in common
+                        prev_cal.map_or(false, |prev_cal| next_cal.map_or(false, |next_cal|prev_cal.overlaps(&next_cal))) &&
+                        // The stay in is not really possible when timing overlaps
+                        // between arrival of first vehicle journey and departure of
+                        // next vehicle journey (see Example 2 above).
+                        last_stop.departure_time <= first_stop.arrival_time
+                    }
+                }
+                _ => false,
+            }
+        };
+
+        self.vehicle_journeys
             .iter()
-            // Filtering out vehicle journeys without any stop time allowing us
-            // below to access first and last stop_times safely.
-            .filter(|(_, vj)| !vj.stop_times.is_empty())
-            .map(|(idx, _)| idx)
-            .collect();
-        // Init all vehicle journeys with:
-        // - no drop-off on first stop
-        // - no pick-up on last stop
+            .filter_map(|(vj_idx, vj)| vj.block_id.clone().map(|b| (b, (vj_idx, vj))))
+            .fold(
+                HashMap::<_, Vec<(Idx<VehicleJourney>, &VehicleJourney)>>::new(),
+                |mut map, (b, (vj_idx, vj))| {
+                    let other_block_id_vj = map.entry(b).or_insert_with(Vec::new);
+
+                    // for every vj we check if it can really be a stay in and if the last stop
+                    // is not in both vj (example 1)
+                    // Note: this is quadratic but should not be too costly since
+                    // the number of vj checked should be limited
+                    other_block_id_vj
+                        .iter()
+                        .for_each(|(other_vj_idx, other_vj)| {
+                            if can_chain_without_overlap(vj, other_vj) {
+                                allowed_first_drop_off_vj.insert(*other_vj_idx);
+                                allowed_last_pick_up_vj.insert(vj_idx);
+                            } else if can_chain_without_overlap(other_vj, vj) {
+                                allowed_first_drop_off_vj.insert(vj_idx);
+                                allowed_last_pick_up_vj.insert(*other_vj_idx);
```

Comment:
Not sure how twisted this example is (all vjs are on the same block_id, same calendar):
vj1 ends at spA (10:00-11:00)
vj2 starts at spA (10:00-11:00) ends at spB (12:00-13:00)
vj3 starts at spC (14:00-15:00)

Because of vj1-vj3 test, the outcome from your code is that all vj1's last dropoff is allowed at spA which is not what we want, no?
(we would have the same for vj2's first pickup if we had a vj0).

This probably implies building a chain first (tree if you consider calendars), then check what's possible... then decide what matters more in case of conflict (probably being safe and forbid).
And it's actually what's done in Navitia as the next vj is unique and check is done once it's affected.

Reply:
hum interesting edge case, I don't know if it's a valid GTFS 🤔
I'm not sure it's worth implementing something this tricky as this is meant only for strengthening the data if we do in `transit_model` I don't think we should be "safe and forbid" (like we should in navitia) since we are only adding additional constraints on the data. So we should just forbid impossible cases, and if a particular case can happen sometimes but not always, we should let it be possible and be handled by the journey planner

Reply:
I think this should be split into 2 parts :
1. Is it a realistic case to have chained blocks (so every extremity would endup being qualified as a valid stay-in in this algo) ? I think we get block_id corresponding to vehicles, so yes (so an algo without determining "real-life"chains would not change data much from original).
2. Is it realistic to have calendars that would make "real-life" chaining of trips being stay-in some days and not stay-in some other days? I think we should not care about that until we have the problem as it's too edgy :see_no_evil: . And OK to stay closer to data in that conflicting case, you make a good point contradicting my "safe" argument.

Reply:
As discussed:
1. Not addressing that for now, so authorize a bit too much pickup/dropoff which is OK (still better to respect data). To be documented but that's all.
2. If calendars introduce a contradictory case (stay-in should be authorized on sunday, not on saturday) : authorise to respect data. To be tested.

Reply:
test added `block_id_on_non_overlaping_calendar_with_overlaping_stops` (and I think the current implementation handles it quite nicely after testing it :stuck_out_tongue_winking_eye: )

Reply:
Indeed, it's nice : what's consistent is handle, and what's not stays close to initial data :+1: 
Letting the discussion unresolved for the "chaining" part if we later come back to that work, but you did all that could reasonably be done :heavy_check_mark: